### ISSUE TITLE
Sql exceptions in async Insight calls are not unwrapped from their aggregate exceptions

### ISSUE THREAD

**Issue by elexisvenator** [jonwagner/Insight.Database#336]
Insight.Database 6.0.0
.Net Framework 4.7
Visual Studio 2017 15.4.1

When using async methods, Sql exceptions thrown by Insight.Database are not thrown the same way as other exceptions.

This can be shown when running unit tests. In the example below I am asserting that an asynchronous call to the database throws an exception. I am using `Microsoft.VisualStudio.TestTools.UnitTesting` as my test framework. Typically when such an assertion is done the test framework can unwrap any aggregate exceptions, but when an Insight Sql exception happens this is not the case.

```c#
[TestClass]
public class SqlExceptionTests
{
        [TestMethod]
        public async Task TestAsyncExceptions()
        {
            // Expects InvalidOperationException, gets InvalidOperationException
            await Assert.ThrowsExceptionAsync<InvalidOperationException>(
                async () =>
                {
                    await Task.Delay(1);
                    throw new InvalidOperationException();
                });

            // Expects SqlException, gets AggregateException
            await Assert.ThrowsExceptionAsync<SqlException>(
                () => CapabilitiDatabase.Connection.ExecuteSqlAsync(
                          "raiserror('Something bad happened', 16, 100);"));
}
```

The same issue can be seen when using the `ExpectedExceptionAttribute`.
This is affecting the ability to write integration tests on some particularly complicated stored procedures.

The exception thrown for a standard async error looks like the following:

```
Capabiliti.FunctionApps.Api.Service.IntegrationTests.Tests.ProcessPeopleTests.FirstNameColumnNotSetReturnsSqlException threw exception: 
System.InvalidOperationException: Operation is not valid due to the current state of the object.
    at Capabiliti.FunctionApps.Api.Service.IntegrationTests.Tests.ProcessPeopleTests.<<FirstNameColumnNotSetReturnsSqlException>g__Func2_0>d.MoveNext() in D:\capabiliti.visualstudio.com\FunctionApps.Api\Capabiliti.FunctionApps.Api.Service.IntegrationTests\Tests\ProcessPeopleTests.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at <NAMESPACE>.SqlExceptionTests.<TestAsyncExceptions>d__2.MoveNext()
```

The exception thrown for the async Insight call looks like the following:

```
TestAsyncExceptions threw exception: 
System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.Data.SqlClient.SqlException: Something bad happened
    at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, String endMethod, Boolean isInternal)
   at System.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at System.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of inner exception stack trace ---
    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Insight.Database.DBConnectionExtensions.<>c__DisplayClass34_0`1.<ExecuteAsyncAndAutoClose>b__2(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of inner exception stack trace ---
    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Insight.Database.DBConnectionExtensions.<>c__DisplayClass34_0`1.<ExecuteAsyncAndAutoClose>b__3(Task`1 t)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at <NAMESPACE>.SqlExceptionTests.<TestAsyncExceptions>d__2.MoveNext()
   ```

**Comment by jonwagner** [comment]
Maybe it's because in the plain case, you're awaiting the result:

                async () =>
                {
                    await Task.Delay(1);
                    throw new InvalidOperationException();
                });

And the insight case, you're returning a task:

                () => CapabilitiDatabase.Connection.ExecuteSqlAsync(
                          "raiserror('Something bad happened', 16, 100);"));

Try this:

            // Expects SqlException, gets AggregateException
            await Assert.ThrowsExceptionAsync<SqlException>(
                async () => await CapabilitiDatabase.Connection.ExecuteSqlAsync(
                          "raiserror('Something bad happened', 16, 100);"));

Or this might work:

            await Assert.ThrowsExceptionAsync<SqlException>(
                           CapabilitiDatabase.Connection.ExecuteSqlAsync(
                          "raiserror('Something bad happened', 16, 100);"));

**Comment by elexisvenator** [comment]
I have created a 2-class test project to highlight the issue.
There are a number of tests in the project to show different ways of triggering the sql exceptions. All other methods I tried worked, except the following:

```c#
[TestMethod]
[ExpectedException(typeof(SqlException))]
public async Task SqlExceptionFromExecuteSqlAsyncThrowsSqlException()
{
    await database.Connection().ExecuteSqlAsync(InvalidSql);
}
```

The test project can be found at https://github.com/elexisvenator/insight-database-async

**Comment by jonwagner** [comment]
I switched all of the old `task.ContinueWith` code with shiny `await` code, so the issue should go away in 6.1.

Of course, that means dropping support for NET35 and NET40, but those old builds are stable anyway.

**Comment by jonwagner** [comment]
This is fixed in 6.1.0-alpha1. (Posting shortly.)

Please let me know if it works for you.

**Comment by elexisvenator** [comment]
I check my projects using the latest alpha, and all the sql exceptions are thrown correctly now. Thanks!
